network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -314,24 +333,42 @@ func (c *ControlMsgValidationInspector) inspectGraftMessages(from peer.ID, graft
 // - prunes: the list of iHaves to inspect.
 // - activeClusterIDS: the list of active cluster ids.
 // Returns:
-//   - DuplicateTopicErr: if there are any duplicate topics found in the list of iHaves
-//     or any duplicate message ids found inside a single iHave.
-//   - error: if any error occurs while sampling or validating topics, all returned errors are benign and should not cause the node to crash.
-//   - bool: true if an error is returned and the topic that failed validation was a cluster prefixed topic, false otherwise.
-func (c *ControlMsgValidationInspector) inspectPruneMessages(from peer.ID, prunes []*pubsub_pb.ControlPrune, activeClusterIDS flow.ChainIDList) (error, p2p.CtrlMsgTopicType) {
+// - p2p.InvCtrlMsgErrs: list of errors that occured during inspection
+func (c *ControlMsgValidationInspector) inspectPruneMessages(from peer.ID, prunes []*pubsub_pb.ControlPrune, activeClusterIDS flow.ChainIDList) p2p.InvCtrlMsgErrs {
+	lg := c.logger.With().
+		Str("peer_id", p2plogging.PeerId(from)).
+		Int("sample_size", len(prunes)).
+		Logger()
 	tracker := make(duplicateStrTracker)
+	errs := make(p2p.InvCtrlMsgErrs, 0)
 	for _, prune := range prunes {
 		topic := channels.Topic(prune.GetTopicID())
 		if tracker.isDuplicate(topic.String()) {
-			return NewDuplicateTopicErr(topic.String(), p2pmsg.CtrlMsgPrune), p2p.CtrlMsgNonClusterTopicType
+			errs = append(errs, p2p.NewInvCtrlMsgErr(NewDuplicateTopicErr(topic.String(), p2pmsg.CtrlMsgGraft), p2p.ModerateErrSeverity, p2p.CtrlMsgNonClusterTopicType))
+			continue
 		}
 		tracker.set(topic.String())
-		err, ctrlMsgType := c.validateTopic(from, topic, activeClusterIDS)
+		err, topicType := c.validateTopic(from, topic, activeClusterIDS)
 		if err != nil {
-			return err, ctrlMsgType
+			// short circuit validation when unstaked peer detected
+			if IsErrUnstakedPeer(err) {
+				lg.Debug().Msg("received control message from unstaked peer")
+				return nil
+			}
```

Comment:
⚠️ it seems this part needs reconsideration; if we keep _continuing_ the for loop, we never report the errors for unstated peers (on cluster prefix channels). Is it the desirable behavior?

Reply:
https://github.com/onflow/flow-go/pull/4978/commits/8a28bf97bad6e97fb921b27887f94b315c26d5f4